Add `getCanonicalTypeName()` for Field and Variable types.

Flone-dnb opened the pull request:
Hello. As I described in one of my issues I needed something like this, so I've spend some time to dig into the source code and I think I'm close to implementing this feature.

This PR adds `getCanonicalTypeName()` for `VariableBase` class which means that both `Field` and `Variable` will have this function. Canonical type name is taken from `kodgen::TypeInfo` class and added to generated files. Since non reflected types can be considered equal (while in reality it may be incorrect) I've put canonical type name to `VariableBase` instead of `Type`.

Here is an example:

```C++
#pragma once

#include <string>
#include <vector>

#include "Foo.generated.h"

using namespace std;

class CLASS() Foo{
public:
    Foo() = default;
    virtual ~Foo() = default;

    FIELD()
    const vector<string> myValue;

    FIELD()
    string myString;

    Foo_GENERATED
};
File_Foo_GENERATED
```

For `myValue` this function will return `std::vector<std::basic_string<char>>` and for `myString` it will be `std::basic_string<char>`.

Due to the lack of knowledge of this project I was unable to compile `RefurekuTests` target, got lots of errors like:

```
TestClass.rfks.h(23, 232): [C2259] 'UniqueInheritedProperty': cannot instantiate abstract class (compiling source file D:\Downloads\Refureku\Refureku\Library\Tests\Src\TestClass.cpp)
```

I've regenerated reflection for test files but was still getting these errors. Maybe you need to regenerate reflection in some special way?

Also, I've noticed that if you don't include `<vector>` type name for `myValue` will not be `std::vector<std::basic_string<char>>` but `int` and the same for string field if you don't include `<string>`. So when `shouldLogDiagnostic` is `true` we can see:

```
...
[Warning] ...h:16:18: error: use of undeclared identifier 'string'
[Warning] ...h:19:5: error: unknown type name 'string'
[Warning] ...h:21:5: error: unknown type name 'Node_GENERATED'
...
```

I think we should detect actual errors (like `use of undeclared identifier 'string'`) and stop code generation and ignore errors like `unknown type name 'Node_GENERATED'`. For this I'm proposing [this PR](https://github.com/jsoysouvanh/Kodgen/pull/4).

jsoysouvanh committed:
[Update] REFUREKU_VERSION macros (2.1.2 -> 2.2.0)

Flone-dnb committed:
Add getCanonicalTypeName() for Field and Variable types.

Flone-dnb committed:
Don't remove namespaces from types.

jsoysouvanh commented:
Hello.

Thank you for your contribution, I'm a bit busy right now but will look into it ASAP.

Flone-dnb committed:
Fix compilation on tests.

Flone-dnb commented:
I've fixed tests target compilation, I had to remove Generated directory so that it will be regenerated using the new generator.

Flone-dnb committed:
Return 1 if failed to generate reflection.

jsoysouvanh commented on Refureku/Library/Include/Internal/Refureku/TypeInfo/Archetypes/StructImpl.h:
```diff
@@ -170,18 +170,21 @@ namespace rfk
+			*	@param name				 Name of the field.
+			*	@param id				 Unique entity id of the field.
+			*	@param canonicalTypeName Name of this variable's type simplified by unwinding all aliases, typedefs,
+			* without qualifiers, namespaces and nested classes.
```
For project nomenclature conformity, please make sure that multi-line parameter documentation is an indentation level higher than the beginning of the first line. See `rfk::Struct::getFieldByName` minFlags parameter for reference.

jsoysouvanh commented on Refureku/Library/Include/Internal/Refureku/TypeInfo/Archetypes/StructImpl.h:
```diff
@@ -191,26 +194,30 @@ namespace rfk
+			*	@param name				 Name of the static field.
+			*	@param id				 Unique entity id of the static field.
+			*	@param canonicalTypeName Name of this variable's type simplified by unwinding all aliases, typedefs,
+			* without qualifiers, namespaces and nested classes.
```
Same as above comment.

jsoysouvanh commented on Refureku/Library/Include/Internal/Refureku/TypeInfo/Archetypes/StructImpl.h:
```diff
@@ -191,26 +194,30 @@ namespace rfk
 			*/
 			RFK_NODISCARD inline StaticField*			addStaticField(char const*		name,
 																	   std::size_t		id,
+																	   const char*      canonicalTypeName,
```
The project uses east const nomenclature so change to `char const*` for consistency.

jsoysouvanh commented on Refureku/Library/Include/Internal/Refureku/TypeInfo/Archetypes/StructImpl.h:
```diff
@@ -191,26 +194,30 @@ namespace rfk
 																	   Struct const*	outerEntity)					noexcept;
 			RFK_NODISCARD inline StaticField*			addStaticField(char const*		name,
 																	   std::size_t		id,
+																	   const char*      canonicalTypeName,
```
The project uses east const nomenclature so change to `char const*` for consistency.

jsoysouvanh commented on Refureku/Library/Include/Internal/Refureku/TypeInfo/Archetypes/StructImpl.inl:
```diff
@@ -49,34 +49,34 @@ inline void Struct::StructImpl::addNestedArchetype(Archetype const* nestedArchet
 }
 
-inline Field* Struct::StructImpl::addField(char const* name, std::size_t id, Type const& type, EFieldFlags flags, 
+inline Field* Struct::StructImpl::addField(char const* name, std::size_t id, const char* canonicalTypeName, Type const& type, EFieldFlags flags, 
```
The project uses east const nomenclature so change to `char const*` for consistency.

jsoysouvanh commented on Refureku/Library/Include/Internal/Refureku/TypeInfo/Archetypes/StructImpl.inl:
```diff
@@ -49,34 +49,34 @@ inline void Struct::StructImpl::addNestedArchetype(Archetype const* nestedArchet
 }
 
-inline StaticField* Struct::StructImpl::addStaticField(char const* name, std::size_t id, Type const& type, EFieldFlags flags, 
+inline StaticField* Struct::StructImpl::addStaticField(char const* name, std::size_t id, const char* canonicalTypeName, Type const& type, EFieldFlags flags, 
```
The project uses east const nomenclature so change to `char const*` for consistency.

jsoysouvanh commented on Refureku/Library/Include/Internal/Refureku/TypeInfo/Archetypes/StructImpl.inl:
```diff
@@ -49,34 +49,34 @@ inline void Struct::StructImpl::addNestedArchetype(Archetype const* nestedArchet
 }
 
-inline StaticField* Struct::StructImpl::addStaticField(char const* name, std::size_t id, Type const& type, EFieldFlags flags, 
+inline StaticField* Struct::StructImpl::addStaticField(char const* name, std::size_t id, const char* canonicalTypeName, Type const& type, EFieldFlags flags, 
```
The project uses east const nomenclature so change to `char const*` for consistency.

jsoysouvanh commented on Refureku/Library/Include/Internal/Refureku/TypeInfo/Variables/VariableBaseImpl.h:
```diff
@@ -31,6 +37,13 @@ namespace rfk
+			* 
+			*	@return _canonicalTypeName.
+			*/
+			inline const char* getCanonicalTypeName() const noexcept;
```
The project uses east const nomenclature so change to `char const*` for consistency.

jsoysouvanh commented on Refureku/Library/Include/Internal/Refureku/TypeInfo/Variables/VariableBaseImpl.h:
```diff
@@ -31,6 +37,13 @@ namespace rfk
 			*	@return _type.
 			*/
 			inline Type const& getType() const noexcept;
```
Please align this method const and noexcept keywords with the below `getCanonicalTypeName` method const and noexcept keywords using tabs.

jsoysouvanh commented on Refureku/Library/Include/Internal/Refureku/TypeInfo/Variables/VariableBaseImpl.inl:
```diff
@@ -5,13 +5,18 @@
+inline VariableBase::VariableBaseImpl::VariableBaseImpl(char const* name, std::size_t id, EEntityKind kind, char const* canonicalTypeName, Type const& type, Entity const* outerEntity) noexcept:
 	EntityImpl(name, id, kind, outerEntity),
-	_type{type}
+	_type{type}, _canonicalTypeName{canonicalTypeName}
```
The project norm enforces only one field initialization per line, so please initialize _canonicalTypeName on a new line.

jsoysouvanh commented on Refureku/Library/Include/Internal/Refureku/TypeInfo/Variables/VariableBaseImpl.inl:
```diff
@@ -5,13 +5,18 @@
 	return _type;
+}
+
+inline const char* VariableBase::VariableBaseImpl::getCanonicalTypeName() const noexcept
```
The project uses east const nomenclature so change to `char const*` for consistency.

jsoysouvanh commented on Refureku/Library/Include/Public/Refureku/TypeInfo/Variables/VariableBase.h:
```diff
@@ -26,6 +26,13 @@ namespace rfk
+			 *
+			 *	@return The name of the variable's type simplified by unwinding all aliases, typedefs.
+			 */
+			REFUREKU_API char const* getCanonicalTypeName() const noexcept;
```
Please align this method const and noexcept keywords with the above `getType` method const and noexcept keywords using tabs.

jsoysouvanh reviewed (changes requested):
Thank you for your contribution.
I have addressed a few comments for the nomenclature consistency.

I still have mixed feelings about having the canonical **type** name at the variable level to be honest though.
From a purely logical point of view, the data should be at the Type level, but in that case the canonical name may or may not be filled: if the type is used for a reflected field/variable or method/function parameter/returned type within the same translation unit, the type canonical name will be valid, otherwise it will stay empty (and let's be honest, this is a huge problem).

I do not have any better idea for now to have this information consistently at the Type level, but I will think about it.

By the way, is it possible for you to modify this pull request to target the dev branch instead of master? I usually have everything on dev and merge to master only for new releases, plus it allows to have the CI run before moving to master.

Flone-dnb commented:
The whole point of this PR is do get type information for non-reflected types, but because `Type` can incorrect for non-reflected types (for example `match`ing `std::string` and `std::vector` will be `true`) I'm adding this functionality to variables instead of types. Although we can 100% work on improving this and eventually move this to the `Type` but right now this will solve my issue and allow me to continue development on my project that uses Refureku.

I will work on requested changes and will change the target branch to `dev`.

Also, because I see lots of code-style mistakes that I made, have you though about using clang-format and/or clang-tidy for code-style? With them developers don't need to care about code-style because clang tools will do all of the formatting automatically (for example, on Ctrl+S press and/or in CI).

Flone-dnb committed:
Apply requested codestyle changes.

Flone-dnb commented:
I consider this PR done. Although I also want you to merge my [other PR](https://github.com/jsoysouvanh/Kodgen/pull/4).

jsoysouvanh commented on Refureku/Library/Include/Public/Refureku/Config.h:
```diff
@@ -9,8 +9,8 @@
 //Version
 #define REFUREKU_VERSION_MAJOR 2
-#define REFUREKU_VERSION_MINOR 1
-#define REFUREKU_VERSION_PATCH 2
```
Version should not change until dev is merged to master for a new library release.

jsoysouvanh commented:
> The whole point of this PR is do get type information for non-reflected types

I get your point here, but it sounds off in the first place wanting to get reflection data from an unreflected class.
If you want to get reflection data, you should reflect your class in the first place, and it's a design decision on your side not wanting to do so.

I will merge this to dev and think about an optimal solution before merging to master. Having these changes pushed to master and then modified afterwards will break the library API/ABI which I want to avoid.
I should be able to work with your forked version until then.

> have you though about using clang-format and/or clang-tidy for code-style?

I know it exists but have never taken the time to look into it yet, will definitely have to do so if it can handle all code-style for us. I wonder how far it can go.

Flone-dnb committed:
Make Object::getArchetype() non pure virtual.

Here is why I've decided to do this:
i've tried creating a custom type T that inherits from `rfk::Object` and
storing it in as a field of some other type. This caused a parsing
error saying that T is abstract. I've noticed that `rfk::Object`'s
`getArchetype()` function is pure virtual and it's implemented
in the generated files, but since on parsing step we don't have
anything generated this causes the parser to show this error. To avoid
this issue I've added a default implementation for `getArchetype()`
that just returns a class with the name `NULL` and size of `int`. Since
all generated files implement `getArchetype()` I don't think this
commit will make any difference to other code.

Flone-dnb committed:
Add missing includes for some tests.

Running the new parsing that fails on parsing errors, some test
files don't include stuff that they use.

Flone-dnb committed:
Add more missing includes for some tests.

Running the new parsing that fails on parsing errors, some test
files don't include stuff that they use.

Flone-dnb committed:
Add shouldFailCodeGenerationOnClangErrors to toml.

I've added the `shouldFailCodeGenerationOnClangErrors` parameter
to the `RefurekuTestsSettings.toml` with the value `false`.

Flone-dnb committed:
Add shouldFailCodeGenerationOnClangErrors to toml.

I've added the `shouldFailCodeGenerationOnClangErrors` parameter to
the `RefurekuSettings.toml` file with the value `false`.

Flone-dnb committed:
Return version to 2.1.2.
